Migrate Jetty Proxy to Airlift (#382)
trinodb/trino-gateway: docs/design.md

# Design

> Trino Gateway is composed of the following main components:

Reviewer: Getting rid of this is good imho .. but we need some more high level design info as well. We can add later though.

other is a request that needs to be forwarded to Trino.

> ## Request Forwarded to Trino

Reviewer: empty line after title

> ## Request Forwarded to Trino

Reviewer: Not sure this is the right document for the routing logic.. but for now we can have it here and then add more info in a follow up PR.

> ## Request Forwarded to Trino

Reviewer: Maybe `HTTP request forwarding to Trino` or jsut `Request forwarding`

> "There are some pre-defined URIs that will be forwarded to Trino. Additional

Reviewer: ```suggestion
Trino Gateway forwards some pre-defined URIs automatically to Trino. You can configure additional
```

> URIs that need to be forwarded are configurable using `extraWhitelistPaths`.

Reviewer: ```suggestion
URIs to forward with the `extraWhitelistPaths` configuration.
```

> In order to support additional URIs that are only known at runtime, we use

Reviewer: Dont use "we"

`RouterPreMatchContainerRequestFilter` to process every request before the
actual resource matching occurs. If the requests URI matches, the request

> will be forwarded to `RouteToBackendResource`.

Reviewer: ```suggestion
is forwarded to `RouteToBackendResource`.
```

> Flow of request forwarding:

Reviewer: ```suggestion
Flow of request forwarding:

```

> 1. Determine which backend Trino to route to.

Reviewer: ```suggestion
1. Determine to which Trino cluster a query should be routed to.
```

> 2. Prepare a request to be sent to Trino. Via header and X-Forwarded headers

Reviewer: ```suggestion
2. Prepare a request to send to Trino by adding header and X-Forwarded headers.
```

>    are being added. Most of the headers are being forwarded to Trino as is.

Reviewer: ```suggestion
   Most headers are forwarded to Trino unmodified.
```

> 3. Depends on the request URI, some of them require special handling. For

Reviewer: ```suggestion
3. Some request URI require special handling. For example, a 
```

>    request which submit a new query, we need to retrieve the queryId from the

Reviewer: ```suggestion
   request which submit a new query, Trino Gateway retrieves the queryId from the
```

>    response from Trino. Some requests to WebUI require setting a session

Reviewer: ```suggestion
   response from Trino. Some requests to the web UI require setting a session
```

   cookie to ensure OIDC works. These are done by chaining asynchronous

>    operations using Future.

Reviewer: ```suggestion
   operations using `Future`.
```

> 4. The execution of request to Trino and the response to client are handled by

Reviewer: ```suggestion
4. The execution of requests to Trino and the response to the client are handled by
```
